### ISSUE TITLE
Support custom design tokens on ThemeData's TextTheme and ColorScheme

### ISSUE THREAD

**Issue by khanhnwin** [flutter/flutter#130918]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Use case

Translating a custom app design in Figma to a Flutter app, the designer provides the foundations for a design system. Functionality, I get everything that I need from the Material library. However, I need to do more work on the widgets to get it looking like the original design. 

The design system has its own design tokens with specifications that the designer has come up with, for example with TextStyles in TextTheme: 
- Heading1
- Heading2
- Heading3
- Subheading1
- Subheading2
- Paragraph
- Label
- etc.

However, these design tokens don't match up with Material's default tokens:
- displayLarge
- displayMedium
- displaySmall
- headlineLarge
- headlineMedium
- headlineSmall
- etc.

Same concept applies for ColorScheme with:
primary, secondary, tertiary, scrim, background, etc. 

Problem: Custom design token names don't match 1:1 with Material's design tokens. 
Solution options:
1. Map custom design tokens to a pre-existing Material's tokens
2. Override the TextStyle or Color directly within the widget, without referencing Material. Likely, creating a standalone class which has the token values. 
3. Override _all_ of the text and colors within the app to one common look and then override individual components to stylize. 

These potential solutions require additional overhead from the developer, whether that be keeping track of two sets of tokens-  the original design tokens and material's, and how they translate -- or overriding all of the styles individually to get it to a good starting point before customizing. It adds friction to the the development flow. 

### Proposal

Allow TextTheme and ColorScheme to be extended with new token names in addition to Material's. Use default value for Material tokens. Perhaps something like... 

```
ColorScheme.custom(
  brand: Color(0xffA2E3F6),
  accent: Color(0xffFFABC7),
)
```

```
TextTheme.copyWithCustom(
  headline1: TextStyle(...),
  subtitle: TextStyle(...),
  label: TextStyle(...),
)
```

**Comment by khanhnwin** [comment]
cc @leighajarett

**Comment by HansMuller** [comment]
You can add colors and text styles to the app's overall theme with [ThemeData.extensions](https://api.flutter.dev/flutter/material/ThemeData/extensions.html). Would that work for your case?

**Comment by leighajarett** [comment]
Yes that works but the problem is it doesn't influence the Material widgets themselves. I don't know how we would even know how to map Material theme text styles to custom design systems without some manual mapping (which you can do today by just overwriting the textStyles in theming. 

So it's not ideal if you want to use any Material widgets in your app (even when folks are creating custom design systems they still use _some_ material widgets. But I don't see how we could offer a better alternative to be honest -- aside from making it easier to understand which Material text style themes influence which Material widget properties so it's easier for folks to map to their own design system.

**Comment by HansMuller** [comment]
Sorry to belabor the obvious: to customize the appearance of the Material components, you can specify component themes. Component themes can refer to color scheme colors that differ from the defaults, or custom colors.  What's missing in all of this?
